### ISSUE TITLE
Provisioning: cluster-local addresses incorrectly shared as peers

### ISSUE THREAD

**Issue by conorsch** [penumbra-zone/penumbra#1815]
See initial report in https://github.com/penumbra-zone/penumbra/issues/1757#issuecomment-1371492956

```
curl -s https://rpc.testnet-preview.penumbra.zone/net_info | jq -r '.result.peers[].remote_ip'
10.8.151.30
10.8.209.137
```

Those RFC1918 addresses are the Service addresses within the cluster. The relevant provisioning logic is here: https://github.com/penumbra-zone/penumbra/blob/ea41ccaa3ab3e3b2a4bd56a7c16651f3ee821665/deployments/ci.sh#L106-L123 We should alter that logic to ensure that only publicly routable addresses are used.

**Comment by hdevalence** [comment]
I think https://github.com/penumbra-zone/penumbra/pull/1660 was supposed to have addressed this, do we know why it didn't?

**Comment by conorsch** [comment]
I must have broken that functionality while trying to implement support for side-by-side testnet and testnet-preview deployments. There's sufficient context in https://discord.com/channels/824484045370818580/930523634945122314/1044834769592143872, including screenshots, to show that the peering capability was indeed working post-merge. I'll take a careful look at the script changes and make sure to comment the functionality once it's fixed again.

**Comment by conorsch** [comment]
Did some digging, here's what I found. We've observed that we can see local-only IP addresses via the Tendermint RPC when inspecting peer info `remote_ip` attributes:

```
curl -s http://testnet-preview.penumbra.zone:26657/net_info | jq -r '.result.peers[].remote_ip'
10.8.209.137
10.8.151.30
```

From reading the Tendermint docs and source, it looks like the `remote_ip` attribute is not the value we should be inspecting (we added auto-inspection in #1672). The `remote_ip` attribute documents the IP address for the peer from the perspective of the tm instance, meaning its closest hop. Since there can be many layers of routing in between, it's not a valid assumption those values will be publicly routable addresses. Compare:

```
curl -s http://testnet-preview.penumbra.zone:26657/net_info | jq -r '.result.peers[].node_info.listen_addr' 
34.28.180.178:26656
34.70.134.40:26656
```

That's more like it! Now we're requesting the addresses that the peers themselves have declared are publicly accessible. That's a great start. However, it's not quite sufficient for our needs. If I run `pd testnet join testnet-preview.penumbra.zone` locally, then I see a my instance's catch-all address naively interpolated in those results:

```
curl -s http://testnet-preview.penumbra.zone:26657/net_info | jq -r '.result.peers[].node_info.listen_addr' 
34.28.180.178:26656
34.70.134.40:26656
tcp://0.0.0.0:26656
```

If I halt my local instance, that `tcp://0.0.0.0:26656` instance will be removed from the RPC endpoint's results. If we want to retain the auto-peer logic from #1672, then I suggest we:

1. update the RPC call from `net_info` `.result.peers[].remote_ip` -> `.result.peers[].node_info.listen_addr`
2. filter the results to strip out useless values like "0.0.0.0"

That should get us back onto an even keel.

**Comment by conorsch** [comment]
On second thought, assuming participants in the testnet will be using our generated configs, then `listen_addr` will _always_ contain "0.0.0.0", so filtering out that value is the same as filtering out all non-PL potential peers. Let's try filtering out RFC1918 addresses from the `pd testnet join` command, which should serve our needs.

**Comment by hdevalence** [comment]
Looks like we can filter out those addresses, at least for IPV4 addresses, using [`is_private`](https://doc.rust-lang.org/std/net/struct.Ipv4Addr.html#method.is_private).

In the future, we could potentially use [`is_global`](https://doc.rust-lang.org/std/net/enum.IpAddr.html#method.is_global), which is provided for the combined v4/v6 `IpAddr` as well as the specific variants, but it's nightly-only and from the [tracking issue](https://github.com/rust-lang/rust/issues/27709) it doesn't seem likely to be stabilized any time soon.

I think it's OK to filter on a best-effort basis, since if we include an invalid address, the peer just won't be able to connect to it.

**Comment by hdevalence** [comment]
> On second thought, assuming participants in the testnet will be using our generated configs, then `listen_addr` will _always_ contain "0.0.0.0", so filtering out that value is the same as filtering out all non-PL potential peers. Let's try filtering out RFC1918 addresses from the `pd testnet join` command, which should serve our needs.

Ah, wait, is the issue here that in some cases (our k8s validators), the `listen_addr` is the correct address to use over the `remote_addr`, but in other cases (like nodes whose configs are generated with `testnet join`), the `listen_addr` is unset, and the `remote_addr` should be used instead?

**Comment by conorsch** [comment]
There are two relevant Tendermint config items, `laddr` and `external_address`: https://github.com/penumbra-zone/penumbra/blob/16c8dc5c7d3f72fe771b861c24b9cdc120ddee8c/testnets/tm_config_template.toml#L207-L215

Neither of those are the same as `remote_ip`, which Tendermint's perspective on the closest hop. Because we set `laddr` to 0.0.0.0 in our generated configs for the testnet, that means 0.0.0.0 will be shown in the TM API as an `external_addr`. That's no more useful to us than an RFC1918 address. 

So I'm not convinced that scraping the node info is the "right" way to allow peers to build a perspective on the network. Shouldn't we be trusting tm do that for us, via pex/gossip?

**Comment by hdevalence** [comment]
Yes, we should be relying on tm to do pex/gossip, the problem we're trying to solve here is just how to bootstrap it, and we don't want to have every node set up with `testnet join` to connect through a single instance.

Longer-term, we can solve that problem by configuring a seed node, but that's not part of the current deployment setup.

In the short term, I think we should:

- use the `listen_addr` instead of `remote_addr`
- filter out `listen_addr` entries that are set to localhost or private addresses

This way, `testnet join` will learn about at least all the addresses in the cluster, as well as any other nodes that have explicitly set their `listen_addr`.

In the medium term (i.e., beyond the immediate goal of standing up testnets again), we could also change `testnet join` to give a way to set the IP address and insert that into the Tendermint config, along with the moniker.  The current templating code is a bit of a mess, though, so it might be good to switch to using [the `tendermint_config` crate](https://docs.rs/tendermint-config/latest/tendermint_config/).

**Comment by hdevalence** [comment]
Should be closed by #1821
